esphome/components/socket/lwip_raw_tcp_impl.cpp

```diff
@@ -178,26 +178,23 @@ class LWIPRawImpl : public Socket {
       errno = EINVAL;
       return -1;
     }
-    if (*addrlen < sizeof(struct sockaddr_in)) {
-      errno = EINVAL;
-      return -1;
-    }
-    struct sockaddr_in *addr = reinterpret_cast<struct sockaddr_in *>(name);
-    addr->sin_family = AF_INET;
-    *addrlen = addr->sin_len = sizeof(struct sockaddr_in);
-    addr->sin_port = pcb_->remote_port;
-    addr->sin_addr.s_addr = pcb_->remote_ip.addr;
+    this->ip2sockaddr_(&pcb_->local_ip, pcb_->local_port, name, addrlen);
     return 0;
   }
   std::string getpeername() override {
     if (pcb_ == nullptr) {
       errno = ECONNRESET;
       return "";
     }
-    char buffer[24];
-    uint32_t ip4 = pcb_->remote_ip.addr;
-    snprintf(buffer, sizeof(buffer), "%d.%d.%d.%d", (ip4 >> 0) & 0xFF, (ip4 >> 8) & 0xFF, (ip4 >> 16) & 0xFF,
-             (ip4 >> 24) & 0xFF);
+    char buffer[50];
+    if (IP_IS_V4_VAL(pcb_->remote_ip)) {
+      inet_ntoa_r(pcb_->remote_ip, buffer, sizeof(buffer));
+    }
+#if LWIP_IPV6
+    else if (IP_IS_V6_VAL(pcb_->remote_ip)) {
+      inet6_ntoa_r(pcb_->remote_ip, buffer, sizeof(buffer));
+    }
+#endif
     return std::string(buffer);
```

Comment:
Probably a bit theoretical, but we should not leak buffer if not used, e.g.:

```suggestion
    char buffer[50];
    if (IP_IS_V4_VAL(pcb_->remote_ip)) {
      inet_ntoa_r(pcb_->remote_ip, buffer, sizeof(buffer));
      return std::string(buffer);
    }
#if LWIP_IPV6
    else if (IP_IS_V6_VAL(pcb_->remote_ip)) {
      inet6_ntoa_r(pcb_->remote_ip, buffer, sizeof(buffer));
      return std::string(buffer);
    }
#endif
    return nullptr;
```

Reply:
Or maybe an empty string? Thoughts @oxan ?

Reply:
I don't see how this can actually trigger until we get IPv8 (and ESP8266's will be obsolete long before that happens), but it should be an empty string, as `nullptr` won't even compile since we're returning `std::string` by value.

Reply:
Maybe just:

```suggestion
    char buffer[50] = {};
    if (IP_IS_V4_VAL(pcb_->remote_ip)) {
      inet_ntoa_r(pcb_->remote_ip, buffer, sizeof(buffer));
    }
#if LWIP_IPV6
    else if (IP_IS_V6_VAL(pcb_->remote_ip)) {
      inet6_ntoa_r(pcb_->remote_ip, buffer, sizeof(buffer));
    }
#endif
    return std::string(buffer);
```